libraries/AP_HAL/UARTDriver.h

```diff
@@ -124,4 +133,40 @@ class AP_HAL::UARTDriver : public AP_HAL::BetterStream {
       return true if this UART has DMA enabled on both RX and TX
      */
     virtual bool is_dma_enabled() const { return false; }
+     * should be called by the main thread @10Hz.  Handles upkeep
+     * tasks like logging.
+     */
+#if HAL_LOG_UART_STATS
+    static void update_logging_all();
+#endif
+
+protected:
+
+#if HAL_LOG_UART_STATS
+    /*
+     * should be called @10Hz to log UART information to dataflash
+     */
+    void update_logging();
+
+    // logging-related variables
+    struct Stats {
+        HAL_Semaphore sem;
```

Comment:
Why is the semaphore within the struct? This means we are copying the semaphore for as part of copying the struct when we are logging (IE it's just wasted time on copying it)